### PR TITLE
[expo-updates][android] refactor FileDownloader to have instance methods

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Currently, using `Updates.checkForUpdateAsync` or `Updates.fetchUpdateAsync` in Android Expo Go will cause the error added in https://github.com/expo/expo/pull/11976 to be thrown.

The error was added because of https://github.com/expo/expo/pull/11875, which added a cache the OkHttpClient instance used in `FileDownloader`. This requires a `Context` object; the current implementation keeps the static OkHttpClient and assumes that `FileDownloader` will only ever be provided `Context`s that have equal return values of `getCacheDir`.

Unfortunately, I did not account for the `ScopedContext` provided to experiences in Expo Go, which has a different `cacheDir` than the application's root context. In this case, we do want to share the OkHttp cache with the root application (the `ScopedContext`'s cache directory is a subdirectory of the root context cache directory); however, this is not something expo-updates should assume or determine on its own.

I decided the correct approach here is to keep track of an instance of FileDownloader (with a member OkHttpClient) and reduce the amount of static methods in this class. This way, the classes that call into FileDownloader can decide whether or not an instance should be shared.

# How

- Make `FileDownloader`'s `OkHttpClient` a member field rather than a static one, and make all methods that use the client non-static.
- Instantiate a copy of `FileDownloader` in `UpdatesController` (for bare apps) and `ExpoUpdatesAppLoader` (for Expo Go / managed apps) and thread it through the stack.

# Test Plan

Tested manually as part of SDK 41 Updates QA

- [x] Expo Go apps update automatically
- [x] all module methods work in Expo Go

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

# Todo

- [ ] cherry pick to master after landing. (I PR'd this directly to sdk-41 because there are almost certainly some merge conflicts with master right now.)

### CODE REVIEW

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.java [hunk 1]
```diff
@@ -55,28 +55,21 @@
...
-  private static Cache getCache(Context context) {
+  private Cache getCache(Context context) {
     int cacheSize = 50 * 1024 * 1024; // 50 MiB
     return new Cache(getCacheDirectory(context), cacheSize);
   }
 
-  private static File getCacheDirectory(Context context) {
-    return new File(context.getCacheDir(), "okhttp");
+  private File getCacheDirectory(Context context) {
+    return new File(context.getApplicationContext().getCacheDir(), "okhttp");
```
Comment (ide) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.java:68]: Think it's worth adding a comment here to explain why the application context is used. This seems like a question someone would not be able to answer by reading this file.

**Review by ide — APPROVED** [review]
Looks good -- I think this is a good way to structure the code and communicate intent clearly in contrast with removing the assertion and accidentally using the root context's cache directory because of the order in which the code just so happens to run.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
 import expo.modules.updates.manifest.ManifestMetadata;
 
 // this unused import must stay because of versioning
-import expo.modules.updates.UpdatesConfiguration;
+
```
Comment (jkhales) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.java:28]: delete the comment above as well?

**Review by jkhales — APPROVED** [review]
✨

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.java (hunk 2 above)
Comment (esamelson) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.java:28]: nice catch -- this import needs to stay actually, like the comment says. Android Studio must have automatically removed the import. I'll revert the diff

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.java (hunk 1 above)
Comment (esamelson) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.java]: oh, good catch. this was from a previous attempted fix that didn't work. i've reverted this change.

**Pull request merged**